cmd/kubeadm/app/cmd/join.go

```diff
@@ -245,27 +296,181 @@ func NewJoin(cfgPath string, args []string, defaultcfg *kubeadmapiv1alpha3.JoinC
+	return nil
+}
+
+// FetchInitClusterConfiguration reads the master configuration from the kubeadm-admin configMap,
+func (j *Join) FetchInitClusterConfiguration(tlsBootstrapCfg *clientcmdapi.Config) (*kubeadmapi.InitConfiguration, error) {
+	// creates a client to access the cluster using the bootstrap token identity
+	tlsClient, err := kubeconfigutil.ToClientSet(tlsBootstrapCfg)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to access the cluster: %v", err)
+	}
+
+	// Fetches the master configuration from the cluster
+	kubeadmConfig, err := configutil.FetchConfigFromFileOrCluster(tlsClient, os.Stdout, "join", "")
 	if err != nil {
+		return nil, fmt.Errorf("Unable to fetch the kubeadm-config ConfigMap: %v", err)
+	}
+
+	// Converts public API struct to internal API
+	masterConfiguration := &kubeadmapi.InitConfiguration{}
+	kubeadmscheme.Scheme.Convert(kubeadmConfig, masterConfiguration, nil)
+
+	return masterConfiguration, nil
+}
+
+// CheckIfReadyForJoinAsMaster ensures that the cluster is in a state that supports
+// joining an additional master node and if the node is ready to join
+func (j *Join) CheckIfReadyForJoinAsMaster(masterConfiguration *kubeadmapi.InitConfiguration) error {
+	// blocks if the cluster was created without a stable controlplane endpoint
+	if masterConfiguration.API.ControlPlaneEndpoint == "" {
+		return fmt.Errorf("unable to join a new master node on a cluster that doesn't have a stable api.controlPlaneEndpoint address")
+	}
+
+	// blocks if the cluster was created without an external etcd cluster
+	if masterConfiguration.Etcd.External == nil {
+		return fmt.Errorf("unable to join a new master node on a cluster that doesn't use an external etcd")
+	}
+
+	// blocks if control plane is self-hosted
+	if features.Enabled(masterConfiguration.FeatureGates, features.SelfHosting) {
+		return fmt.Errorf("`kubeadm join --experimental-master` is still in alpha and self-hosted clusters are not supported yet")
+	}
+
+	// blocks if the certificates for the control plane are stored in secrets (instead of the local pki folder)
+	if features.Enabled(masterConfiguration.FeatureGates, features.StoreCertsInSecrets) {
+		return fmt.Errorf("`kubeadm join --experimental-master` is still in alpha and certificates stored in secrets are not supported yet")
+	}
+
+	// checks if the certificates that must be equal across masters are provided
+	if ret, err := certsphase.SharedCertificateExists(masterConfiguration); !ret {
 		return err
 	}
 
+	return nil
+}
+
+// PrepareForJoinAsMaster makes all preparation activities require for a node joining as a master
+func (j *Join) PrepareForJoinAsMaster(masterConfiguration *kubeadmapi.InitConfiguration) error {
+
+	// Creates the admin kubeconfig file for the admin and for kubeadm itself.
+	if err := kubeconfigphase.CreateAdminKubeConfigFile(kubeadmconstants.KubernetesDir, masterConfiguration); err != nil {
+		return fmt.Errorf("error generating the admin kubeconfig file: %v", err)
```

Comment:
could you use errors.Wrap https://godoc.org/github.com/pkg/errors

Reply:
We don't use errors.Wrap anywhere else in the kubeadm codebase. Please use fmt.Errorf for consistency.